Adds a spec to check enterprise fees creation

filipefurtad0 opened the pull request:
Prepares for #10348.

#### What? Why?

- Adds a spec to check enterprise fees creation and cover the cases in which we have invalid combinations of tax category and calculator

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Green build after merging and rebasing https://github.com/openfoodfoundation/openfoodnetwork/pull/10512 and #10348 is closed.

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: Technical changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

mkllnk commented on spec/system/admin/enterprise_fees_spec.rb:
```diff
@@ -177,27 +177,56 @@
+        end
+      end
+
+      xcontext "an error message is displayed" do
+        # pending "#10348"
```
Why `x` instead of `pending`?

filipefurtad0 replied:
I could not get it to work with pending, I think this is because pending would need to be inside the `shared_examples` block, which would influence the three other (not pending) test cases.

So, If I remove the `x`, and uncomment the pending line, the spec fails. I thought of adding a condition within the `shared_examples` but it seems like too much hassle...

Any ideas on how to make it work with pending in this case?

mkllnk replied:
Ah, now I see. All good. Shared examples with some pending is not easy to write. You found the simplest solution.

abdellani reviewed (approved):
Looks good to me
thank you @filipefurtad0

mkllnk commented:
Moved to In Dev for rebasing after the other PR is merged.

RachL commented:
@filipefurtad0 is this one still blocked? If so by what?

RachL commented:
Thanks I've added it as a dependencies 👍 1:

filipefurtad0 commented:
Rebased, solved conflicts and added a spec to cover for this [case -> 4.](https://github.com/openfoodfoundation/openfoodnetwork/pull/10512#issuecomment-1531161376) (editing an existing fee into an invalid combination.)

dacook commented on spec/system/admin/enterprise_fees_spec.rb:
```diff
@@ -191,28 +207,60 @@
+                   from: 'sets_enterprise_fee_set_collection_attributes_0_enterprise_id'
+            select 'Packing', from: 'sets_enterprise_fee_set_collection_attributes_0_fee_type'
+            fill_in 'sets_enterprise_fee_set_collection_attributes_0_name', with: 'foo'
+            select tax_category, 
+from: 'sets_enterprise_fee_set_collection_attributes_0_tax_category_id'
```
Funny indenting here. I can see it's due to the challenge of the very long field name.
(Although it's exacerbated by the deep level of indentation too).

I can see this string is repeated a lot: `sets_enterprise_fee_set_collection_attributes_0`. Perhaps we can set that once in advance, then we can do something simpler like

```suggestion
            select tax_category, from: "#{prefix}_tax_category_id"
```

dacook replied:
Good one, that makes sense too.

dacook commented on spec/system/admin/enterprise_fees_spec.rb:
```diff
@@ -191,28 +207,60 @@
+        click_link "Create One Now"
+      end
+
+      shared_examples "shared example" do |tax_category, calculator, flash_message, fee_count|
```
> "shared example"

It would be better to have a meaningful name for this. Perhaps something like "setting tax category with a calculator". Hmm I see the duplication with the below context name, but still I think that's better than nothing.

filipefurtad0 replied:
Nice catch, this one slipped. I think I've left the naming to the end, but probably forgot about it... 

Before it read:

```
As an administrator
    I want to manage enterprise fees
  as an enterprise manager
    creating an enterprise fee
      an error message is displayed
        behaves like shared example
          setting it up as Inherit From Product, with a Flat Rate (per order) calculator
            triggers the expected message
      an success message is displayed
        behaves like shared example
          setting it up as Inherit From Product, with a Flat Rate (per item) calculator
            triggers the expected message
      an success message is displayed
        behaves like shared example
          setting it up as GST, with a Flat Rate (per order) calculator
            triggers the expected message
      an success message is displayed
        behaves like shared example
          setting it up as GST, with a Flat Rate (per item) calculator
            triggers the expected message

```

After the latest commit, it reads:
```

As an administrator
    I want to manage enterprise fees
  as an enterprise manager
    creating an enterprise fee
      an error message is displayed
        behaves like setting it up
          as Inherit From Product, with a Flat Rate (per order) calculator
            triggers the expected message
      an success message is displayed
        behaves like setting it up
          as Inherit From Product, with a Flat Rate (per item) calculator
            triggers the expected message
      an success message is displayed
        behaves like setting it up
          as GST, with a Flat Rate (per order) calculator
            triggers the expected message
      an success message is displayed
        behaves like setting it up
          as GST, with a Flat Rate (per item) calculator
            triggers the expected message
```
Hope this is ok, let me know what you think.

dacook replied:
This makes sense 👍

dacook reviewed (approved):
Great job, and it's satisfying to see the end result in the final commit :D

I have a couple of comments for improvement, so will move it into In Dev. If you haven't got the time to do them though, feel free to merge.

dacook reviewed (approved):
Great job, and it's satisfying to see the end result in the final commit :D

I have a couple of comments for improvement, so will move it into In Dev. If you haven't got the time to do them though, feel free to merge.

filipefurtad0 commented:
Thanks for your review @dacook - I've now made the proposed changes. Moved back to code review.
